tests/pytorch/distributed/run_numerics.py

```diff
@@ -0,0 +1,701 @@
+
+    dist.init_process_group(backend="nccl")
+    NCCL_WORLD = dist.new_group(backend="nccl")
+
+    WORLD_RANK = dist.get_rank()
+    WORLD_SIZE = dist.get_world_size()
+    assert WORLD_SIZE == 2, "This test uses 2 GPUs. Run with torchrun --nproc_per_node=2."
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument("-l", "--layer-type", type=str)
+    parser.add_argument("--fp8", action="store_true", default=False)
+    args = parser.parse_args(argv, namespace)
+
+    test_dict = {
+        "linear": test_linear,
+        "layernorm": test_layernorm,
+        "layernorm_linear": test_layernorm_linear,
+        "layernorm_mlp": test_layernorm_mlp,
+        "transformer_layer": test_transformer_layer,
+    }
+
+    assert args.layer_type in test_dict, f"Argument --layer-type must be in {test_dict.keys()}"
+
+    FP8 = args.fp8
+
+    output_value = test_dict[args.layer_type]()
+    dist.destroy_process_group()
+    return output_value
+
+
+def run_distributed_test(test_name=None):
+    def decorator(func):
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            name = test_name if test_name is not None else func.__name__
+
+            dist_print(f"Starting test {name} with args {args} and {kwargs}")
+            torch.cuda.set_device(WORLD_RANK)
+            torch.manual_seed(1234)
+            torch.cuda.manual_seed(1234)
+            func(*args, **kwargs)
+
+            dist.barrier()
+            dist_print(f"Passed test {name}")
+
+        return wrapper
+
+    return decorator
+
+
+def _gather(tensor, dim=0):
+    """
+    Gathers tensors and concats them. Since torch.distributed.nn.functional.all_gather
+    multiplies gradients by WORLD_SIZE, those gradiedts are rescaled.
+    """
+
+    class HalfGradient(torch.autograd.Function):
+        @staticmethod
+        def forward(ctx, input):
+            ctx.save_for_backward(input)
+            return input  # forward pass (identity)
+
+        @staticmethod
+        def backward(ctx, grad_output):
+            return grad_output / WORLD_SIZE  # gradient division by WORLD_SIZE
+
+    tensor = HalfGradient.apply(tensor)
+    gathered = torch.distributed.nn.functional.all_gather(tensor, group=NCCL_WORLD)
+    return torch.cat(gathered, dim=dim)
+
+
+def _constant(tensor):
+    return nn.init.constant_(tensor, 0.5)
+
+
+def dist_print(msg, src=None, end="\n", error=False):
+    stream = sys.stderr if error else sys.stdout
+    if WORLD_RANK == (0 if src is None else src):
+        stream.write(f"[rank{WORLD_RANK}] {msg}{end}\n")
+    dist.barrier()
+
+
+def _check_outputs(output_single_node, output_distributed):
+    numerics_failed = torch.tensor([0], dtype=torch.uint8, device="cuda")
+    rtol = 0.125 if FP8 else 0.025
+    atol = 0.0625 if FP8 else 0.00125
+    output_failed, output_info = _compare_tensors(
+        "outputs", output_distributed, output_single_node, rtol, atol
+    )
+    dist_print(output_info, src=WORLD_RANK, error=output_failed)
```

Comment:
Printing successful checks might be too verbose:
```suggestion
    if output_failed:
        dist_print(output_info, src=WORLD_RANK, error=True)
```

Reply:
changed